File: backend/src/sidra/api/models/stats_response.py
from __future__ import annotations

from datetime import date

from pydantic import BaseModel

from sidra.api.models.stats_track import StatsTrack


class StatsStanding(BaseModel):
    """The sidra at a glance, counted in tracks rather than units."""

    behind: int
    on_pace: int
    ahead: int
    not_started: int
    chavrusa: int


class StatsStreak(BaseModel):
    """Consecutive days with something recorded, across the whole sidra."""

    current: int
    longest: int


class StatsResponse(BaseModel):
    """Everything the Stats screen draws."""

    on: date
    days: list[date]
    """The window's columns, in order. Never empty."""

    window_days: int
    requested_window_days: int
    """What was asked for. The window is clamped to the ledger's own age, and a report that hides
    the clamp claims a history it does not have."""

    standing: StatsStanding
    streak: StatsStreak
    tracks: list[StatsTrack]

File: frontend/src/types/StatsResponse.ts
export interface StatsTrack {
  readonly track_id: string;
  readonly name_en: string;
  readonly name_he: string;
  readonly unit_singular: string;
  readonly unit_plural: string;
  /** Null on a chavrusa track, which carries staleness rather than debt. */
  readonly debt_now: number | null;
  /** Where the debt stood when the window opened, so the direction of travel is visible. */
  readonly debt_then: number | null;
  readonly learned_units: number;
  readonly days_learned: number;
  readonly last_learned_on: string | null;
  /** First real advance. Null means never opened, which is not the same as opened and idle. */
  readonly opened_on: string | null;
  /** Billed minus learned, one per day. Positive opened the gap, negative closed it. */
  readonly net: readonly number[];
}

export interface StatsStanding {
  readonly behind: number;
  readonly on_pace: number;
  readonly ahead: number;
  readonly not_started: number;
  readonly chavrusa: number;
}

export interface StatsResponse {
  readonly on: string;
  readonly days: readonly string[];
  readonly window_days: number;
  /** What was asked for. The window is clamped to the ledger's own age. */
  readonly requested_window_days: number;
  readonly standing: StatsStanding;
  readonly streak: { readonly current: number; readonly longest: number };
  readonly tracks: readonly StatsTrack[];
}
